lib/reline/line_editor.rb

```diff
@@ -873,46 +874,45 @@ def editing_mode
     end
   end
 
-  private def move_completed_list(list, direction)
-    case @completion_state
-    when CompletionState::NORMAL, CompletionState::COMPLETION,
-         CompletionState::MENU, CompletionState::MENU_WITH_PERFECT_MATCH
-      @completion_state = CompletionState::JOURNEY
-      result = retrieve_completion_block
-      return if result.nil?
-      preposing, target, postposing = result
-      @completion_journey_data = CompletionJourneyData.new(
-        preposing, postposing,
-        [target] + list.select{ |item| item.start_with?(target) }, 0)
-      if @completion_journey_data.list.size == 1
-        @completion_journey_data.pointer = 0
-      else
-        case direction
-        when :up
-          @completion_journey_data.pointer = @completion_journey_data.list.size - 1
-        when :down
-          @completion_journey_data.pointer = 1
-        end
+  def dialog_proc_scope_completion_journey_data
+    return nil unless @completion_journey_state
+    line_index = @completion_journey_state.line_index
+    pre_lines = @buffer_of_lines[0...line_index].map { |line| line + "\n" }
+    post_lines = @buffer_of_lines[(line_index + 1)..-1].map { |line| line + "\n" }
+    DialogProcScope::CompletionJourneyData.new(
+      pre_lines.join + @completion_journey_state.pre,
+      @completion_journey_state.post + post_lines.join,
+      @completion_journey_state.list,
+      @completion_journey_state.pointer
+    )
+  end
+
+  private def move_completed_list(direction)
+    if @completion_journey_state
+      if (delta = { up: -1, down: +1 }[direction])
+        @completion_journey_state.pointer = (@completion_journey_state.pointer + delta) % @completion_journey_state.list.size
       end
-      @completion_state = CompletionState::JOURNEY
     else
-      case direction
-      when :up
-        @completion_journey_data.pointer -= 1
-        if @completion_journey_data.pointer < 0
-          @completion_journey_data.pointer = @completion_journey_data.list.size - 1
-        end
-      when :down
-        @completion_journey_data.pointer += 1
-        if @completion_journey_data.pointer >= @completion_journey_data.list.size
-          @completion_journey_data.pointer = 0
-        end
-      end
+      preposing, target, postposing = retrieve_completion_block
+      return false unless target
+
+      list = call_completion_proc
+      return false unless list.is_a?(Array)
+
+      candidates = list.select{ |item| item.start_with?(target) }
```

Comment:
> Is it possible to create a list that does not start with target?

Yes, it's possible by this code
```ruby
require 'reline'
Reline.autocompletion = true
Reline.completion_proc = ->*{['abc','def','ghi','jkl']}
p Reline.readmultiline('>'){true}
```

I'll remove this `select` because older reline does not reject candidates that does not start with target.
```
> x
  abc
  def
  ghi
  jkl
```

Reply:
The old code was doing this select in L886
```ruby
private def move_completed_list(list, direction)
  ...
  @completion_journey_data = CompletionJourneyData.new(
    preposing, postposing,
    [target] + list.select{ |item| item.start_with?(target) }, 0)
```
The old behavior (renders candidates that does not start from target, but disappears after pressing tab) seems buggy.